src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -42,10 +42,13 @@ def interpreter_constraints(self) -> InterpreterConstraintsField:
 # Normally we would subclass `DeduplicatedCollection`, but we want a custom constructor.
 class InterpreterConstraints(FrozenOrderedSet[Requirement], EngineAwareParameter):
     def __init__(self, constraints: Iterable[str | Requirement] = ()) -> None:
-        super().__init__(
-            v if isinstance(v, Requirement) else self.parse_constraint(v)
-            for v in sorted(constraints, key=lambda c: str(c))
-        )
+        # #12578 `parse_constraint` will sort the requirement's component constraints into a stable form.
+        # We need to sort the component constraints for each requirement _before_ sorting the entire list
+        # for the ordering to be correct.
+        parsed_constraints = [
+            i if isinstance(i, Requirement) else self.parse_constraint(i) for i in constraints
+        ]
```

Comment:
Better is to use a generator expression:

```suggestion
        parsed_constraints = (
            i if isinstance(i, Requirement) else self.parse_constraint(i) for i in constraints
        )
```

You avoid the intermediate data structure.

Reply:
Or you could keep the list and call `.sort(key=lambda c: str(c))` on it.

Reply:
That's a holdover from putting some `print` calls in while I verified behaviour. Thanks for catching this.